NEW blog posts: python environments (#20)
lwasser/lwasser.github.io: _posts/2022-08-15-python-conda-environments-create.md

---

> title: "Setup Python Conda Environments for reproducible science and for teaching Earth and Environmental Data Science"

Author: ```suggestion
title: "Setup Python Conda Environments for open science workflows and for teaching Earth and Environmental Data Science"
```

> excerpt: "Teaching earth and environmental data science can be time consuming. Learn about what I learned when I tried to create an consistent Python environment that was easy for students to install. (and how this could help your science as well."

Author: ```suggestion
excerpt: "Sometimes the biggest time sink in getting started with the Python programming language is creating a working environment with all of the tools that you need. In this blog I will talk about how I created a tested Python conda environment that worked on different operating systems. This made it easier for students and people working through online lessons to setup a consistent Python environment with all of the packages that they needed."
```

layout: single_law
related: true
permalink: tested-python-environments-for-teaching-data-science/
author_profile: true
comments: true
header:
  overlay_color: "#333"
series: "conda-envt-testing"
category: "open-science"
categories:
  - testing
  - python
  - open-source
  - open-education
---
<!--<original permalikn free-open-source-software-for-science-foss/> -->

## Why should scientists care about environments and testing things?

Environments and testing are critical when you are thinking about both 
teaching earth and environmental data science or just doing research with the 

> goal of making your work open and reproducible. When I saw open and reproducible, i'm referring to a workflow that others can use and build upon. Vs having to start over from scratch to recreate your work.  

Author: ```suggestion
goal of making your work open and reproducible. When I say open and reproducible, i'm referring to a workflow that others can use and build upon. The alternative is that anyone who wants to run needs to build their own from scratch. That definitely is not the spirit of open science!  
```

> When I was just starting out in science, if you had suggested the idea of testing and working on environments
> as being a part of my workflows I would have laughed. In fact, if you told me 
> id be teaching data science 20 years later and running an organization devoted
> to the tools that support data science, I'd laugh even harder. 

Author: ```suggestion
When I was a young scientist, I would never have considered reusable environments and testing as a part of the skillset that I needed.  In fact, if you told me 
that i'd be teaching data science 20 years later and running an organization devoted
to the tools that support data science, I would have laughed. 
```

> But testing has become and important part of any infrastructure that I create 

Author: ```suggestion
But testing has become an important part of any infrastructure that I create 
```

be it:

* documentation

> * environments
> * or tools

Author: ```suggestion
* environments
* tools
* or even courses that I teach or online lessons that I create
```

> I'll devote another blog to setting up testing. 

Author: ```suggestion
```

> This blog is about how to setup a consistent and easier-to-install Python (conda) environment. Setting up consistent environments can help you create better,
> easier-to-reproduce workflows. This is particularly important if you are trying
> to help students setup their computers and create workflows in Python. But it 
> will also support your scientific workflows!

Author: ```suggestion
In this blog, I'll discuss how to setup a consistent and easier-to-install Python (conda) environment. Setting up consistent environments can help you create better,
easier-to-reproduce workflows. This is particularly important if you are trying
to help students setup their computers and create workflows in Python. It also supports creating open scientific workflows.
```

> ## My Path From Using Matlab to Teaching R & Python

Author: ```suggestion
## The backstory: My Path From Using Matlab to Teaching R & Python
```

I could write vignettes for packaging. I could easily create reproducible 
reports with citations (`RMarkdown`).

Using R, I could run a workshop and send out setup instructions. People 
installed the packages and came ready to learn. Just like that. Out of the 
imaginary "box". Sure `GDAL` was sometimes problematic in those year working with
R. But I figured it out. 

*(Does anyone actually put software in boxes anymore?)*

## Here Comes Python... and GDAL - Help!!

My move to the Python programming language was much less graceful. I moved because 
my program at CU Boulder was rooted in the Earth sciences. The earth science 
community was using Python more regularly than R. I determined this by surveying 
the community - public and private sector. 

My surveys told me that industry was using `Python` a lot too. It's 
generic design lends it to be used in may different fields and applications. 
I knew that students could benefit professionally across many fields if they knew
how to code in `Python`. 

*By the way, I still think R is great for many things!
I still use R too! No judgement here!*
{: .notice }

### How Does One Install GDAL Anyway?

> But, installing Python packages was much less intuitive. Quickly I had to figure 
> out whether to use `pip` or `conda` to manage installing package. And once I 
> settled on conda, it became what channel do I use? `conda-forge` seemed like 
> a good one. I found myself with environment
> conflicts. 

Author: ```suggestion
But, installing Python packages was much less intuitive than it was in R. Quickly I had to figure 
out whether to use `pip` or `conda` to manage installations. And once I 
settled on `conda`, it became what channel do I use? `conda-forge` seemed like 
a good one. But alas, I found myself with environment
conflicts. 
```

> oh yes, conflict.

Author: ```suggestion
```

> I'd install one thing and it would have a conflict with something else. Id remove that
> other thing and something else would break. 

Author: ```suggestion
I'd install one tool and it would have a conflict with something else. I'd remove that
tool causing the conflict and something else would break. And so it went. :sob:
```

> <figure>
> 	<a href="/images/xkcd-python-environment.png">
>     <img src="/images/xkcd-python-environment.png"></a>
> 	<figcaption>Environment pain when working with Python is real. However, Python
>   is an awesome programming language that is broadly used across disciplines. So 
>   it's worth figuring this all out. <i>Source xkcd.</i> 
>   </figcaption>
> </figure> 

Author: ```suggestion
<figure>
	<a href="/images/xkcd-python-environment.png">
    <img src="/images/xkcd-python-environment.png"></a>
	<figcaption>Behold, the environment spaghetti monster. Environment pain when working with Python is real. But there are solutions to this in the form of environments. Ideally tested environments. <i>Source xkcd.</i> 
  </figcaption>
</figure> 
```

> 
> I  just wanted to code. And teach others how to code.
> 

Author: ```suggestion
```

> I found myself digging around for help and trying to interpret heavily 

Author: This entire section could be tightened up... just a few sentences get to the point

	<figcaption>Meme created by a colleague, Filipe about installing GDAL in a 
  Python environment. It just is never pretty, friends. Now imagine students who 
  are new to python and want to get setup for spatial data trying just to install
  things... ouch!  
  </figcaption>
</figure>

## I just wanted to teach people how to process data in a reproducible way

I didn't want to students to spend weeks trying to get an environment working.
That is how you lose people.

> If I can't install software, how can I code?
> -- Says every student ever who deals with imposter syndrome

So what was my next step?

## I (we) built the Earth Analytics Python Environment (& setup testing)

I knew that having an easier-to-install / almost fail proof Python environment 
would make teaching easier.
And better yet, if I could test that environment on Linux, Mac and Windows 
(students use all sorts of computers and OS' - you can't assume only one) then
a lot of my issues would be solved. 
I would know before they installed the environment where the pain points were. I could troubleshoot before class rather than during class. 

This was huge!

### Enter the Earth Analytics Python Environment

> So this is what I built:

Author: ```suggestion
So this is what I built *(with the help of a few amazing colleagues)*:
```

> <i class="fas fa-hands-helping"></i> Thanks to the help of several people 
> including a great colleague 
> [Max Joseph](https://www.github.com/mbjoseph/). Max was really the person who
> early on introduced me to the concept of testing. At the time I didn't fully
> understand just how important it was. But I got there quickly!
> {: .notice }

Author: ```suggestion
```

> 1. [A vetted Python environment - this links to the original repo on EarthLab's organization. Below I link to a fork in case that repo changes over time](https://www.github.com/earthlab/earth-analytics-python-env/): This environment has all of the Python tools needed to complete my earth-analytics-python class. Tools for remote sensing and other spatial data. Tools for natural language processing, etc. And it can be installed quickly and easily using a recipe file (also known as an environment file - more on that below).

Author: ```suggestion
1. [A vetted Python environment - I forked my original repo in Earth Lab which has many more stars as it doesn't seem to be maintained anymore](https://www.github.com/lwasser/earth-analytics-python-env/): This environment has all of the Python tools needed to complete my earth-analytics-python class. Tools for remote sensing and other spatial data. Tools for natural language processing, etc. And it can be installed quickly and easily using a recipe file (also known as an environment file - more on that below).
```

2. The environment has a [GitHub Actions workflow that tested it on Windows, Mac and Linux.](https://github.com/lwasser/earth-analytics-python-env/tree/main/.github/workflows). I'll talk about that part more in a second, followup blog. 

*The original workflow was built using Travis CI and Appveyor - yikes!*

<i class="fas fa-hands-helping"></i> Thanks to [Filipe's help](https://www.github.com/ocefpaf), in a separate 
environment that I created for an automated build, we used Conda Lock. We did 
this so we could pin down exactly what versions of all tools and dependencies 
were being installed once we have a working environment. Boy this made 
troubleshooting easier! More on that in another blog.
{: .notice}

## What do all of the pieces in this earth-analytics-python repository do?

> Ok so let's break down what is in this [`earth-analytics-python` GitHub repo](https://www.github.com/lwasser/earth-analytics-python-env):
> 

Author: ```suggestion
Next, i'll break down the key elements of this environment, found in the [`earth-analytics-python` GitHub repo](https://www.github.com/lwasser/earth-analytics-python-env):

```

> ### First you have the conda Python environment file. What's that? 

Author: ```suggestion
### First you have the conda Python environment.yml file. What's that? 
```

> You can think of a `Python` environment file, as a recipe for

Author: I think the toolbox explanation needs to go up here beore the file runt hrough below...

> You can think of a `Python` environment file, as a recipe for
> what to install and where. By using this environment file, 
> we are making a decision now to use the `conda` package manager 
> to install packages vs `pip`. 

Author: ```suggestion
You can think of a `Python` environment file, as a recipe for
what to install and where. The environment file, 
tells the installer how to install each tool.  It can also be used to specify specific tool versions. And you can tell it what package manager and channel to use:
* `pip` vs `conda`,
* `conda-forge` vs `defaults`,
etc. 
```

<i class="fas fa-info-circle"></i> The package manager just helps

> keep track of what you install and helps you decide what toolbox 

Author: where did i add the toolbox explanation of channels? did i miss that section ?

> <i class="fas fa-info-circle"></i> The package manager just helps
> keep track of what you install and helps you decide what toolbox 
> to install that package from. It also any dependencies that
> the packages you are installing has and addresses potential 
> conflicts with different versions of different dependencies.
> {: .notice } 

Author: ```suggestion
<i class="fas fa-info-circle"></i> The package manager just helps
keep track of what you install. It also allows you to specify what toolbox 
to install that package from. It also any dependencies that
the packages you are installing has and addresses potential 
conflicts with different versions of different dependencies.
{: .notice } 
```
